libyul/backends/wasm/EVMToEWasmTranslator.cpp

```diff
@@ -0,0 +1,370 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Translates Yul code from EVM dialect to eWasm dialect.
+ */
+
+#include <libyul/backends/wasm/EVMToEWasmTranslator.h>
+
+#include <libyul/backends/wasm/WordSizeTransform.h>
+#include <libyul/backends/wasm/WasmDialect.h>
+#include <libyul/optimiser/ExpressionSplitter.h>
+#include <libyul/optimiser/FunctionGrouper.h>
+#include <libyul/optimiser/MainFunction.h>
+#include <libyul/optimiser/FunctionHoister.h>
+#include <libyul/optimiser/Disambiguator.h>
+#include <libyul/optimiser/NameDisplacer.h>
+
+#include <libyul/AsmParser.h>
+#include <libyul/AsmAnalysis.h>
+#include <libyul/AsmAnalysisInfo.h>
+#include <libyul/Object.h>
+
+#include <liblangutil/ErrorReporter.h>
+#include <liblangutil/Scanner.h>
+#include <liblangutil/SourceReferenceFormatter.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+using namespace langutil;
+
+namespace
+{
+static string const polyfill{R"({
+function or_bool(a, b, c, d) -> r {
+	r := i64.ne(0, i64.or(i64.or(a, b), i64.or(c, d)))
+}
+function add_carry(x, y, c) -> r, r_c {
```

Comment:
Perhaps speciy the domain? What is the acceptable range of `c`

Reply:
What happens when `x+y+c >= 2 * (2^64)`?

Reply:
Ah you are right, `c` can be at most 1 in which case it is not useable for `sub` anymore. :(